plugins/modules/firewall.py

```diff
@@ -242,23 +242,57 @@ def _normalize_ips(rules: list) -> list:
 
         return result
 
-    def _rules_changed(self) -> bool:
-        """Returns whether the local and remote firewall rules match."""
+    def _amend_rules(self, remote_rules: list, local_rules: list) -> None:
+        # produce a result list in the same order as remote_rules
+        # amended by the updates from the local_rules
+
+        local_labeled_rules = { r['label']:r for r in local_rules }
+        result=[]
+        for remote_rule in remote_rules:
+            # copy remote_rule as is if not being updated by local_rules
+            if remote_rule['label'] not in local_labeled_rules:
+                result.append(remote_rule)
+                continue
+
+            # insert all missing fields in local_rule from remote_rule
+            local_rule = local_labeled_rules[remote_rule['label']]
+            for field in linode_firewall_rule_spec.keys():
+                if field not in local_rule:
+                    local_rule[field]=remote_rule[field]
+            result.append(local_rule)
+        return result
+
+    def _update_rules(self) -> list:
+        """Updates remote firewall rules relative to user-supplied new rules, and returns whether anything changed."""
         local_rules = filter_null_values_recursive(self.module.params['rules'])
+        # user did not specify any rules updates
+        if local_rules == None: return []
+
         remote_rules = filter_null_values_recursive(mapping_to_dict(self._firewall.rules))
 
         # Normalize IP addresses for all rules
-        for field in {'inbound', 'outbound'}:
-            if field in local_rules:
-                local_rules[field] = self._normalize_ips(local_rules[field])
-            else:
-                # We should normalize missing keys to [] for diffing purposes
-                local_rules[field] = []
-
-            if field in remote_rules:
-                remote_rules[field] = self._normalize_ips(remote_rules[field])
-
-        return local_rules != remote_rules
+        for direction in ['inbound', 'outbound']:
+            local_rules[direction] = self._normalize_ips(local_rules[direction]) if direction in local_rules else []
+            remote_rules[direction] = self._normalize_ips(remote_rules[direction]) if direction in remote_rules else []
+
+        if self._state == "update":
```

Comment:
I wonder if there's a way we could break this out into another function or use guard clauses to make this a bit more readable. No big deal either way since it's primarily a preference and formatting thing :slightly_smiling_face:

Reply:
I opted not to do that for now.